File: webapp/js/dockerimagedeploy.js
if(ws==undefined)var ws;
if(ws)ws.close();
ws = new WebSocket("ws://localhost:" + global.backendPort + "/web/app/events");
//通信が接続された場合
ws.onopen = function(e) { 

    console.log("Setup web socket:", ws);
    ws.send(JSON.stringify({
        "event": "dockerimagedeploy",
    }))  
    ws.onmessage = function (event) {
        
        const eObj = JSON.parse(event.data);
        console.log(eObj);
        const data = JSON.parse(eObj.event);
        if(data.eventName == "index-view") {
            $(".content-wrapper").html(NNSH_decodeHTML(data.view.view_data));
            return
        }
        else if(data.eventName == "log") {
            document.getElementById("console-output").innerHTML += "\n" + data.log;
            $('#console-output').animate({scrollTop: $('#console-output')[0].scrollHeight}, 'fast');
            return
        }
        if(data.eventName == "dockerimagedeploy-saveinput") {
            if(data.deployInputlist) {
                try {$('#inputlist').columns('destroy');} catch(e) {}
                $('#inputlist').columns({data:data.deployInputlist,
                    schema: [
                        {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadDeployInput(\'{{name}}\')">{{name}}</a>'},
                        {"header":"", "key":"cmd"},
                        {"header":" ", "key":"", "template":'<input type="button" onclick="deleteDeployInput(\'{{name}}\')" value="DELETE" />'},
                ]
                });
            }
            return
        }
        if(data.eventName == "dockerimagedeploy-delete") {
            if(data.deployInputlist) {
                try {$('#inputlist').columns('destroy');} catch(e) {}
                $('#inputlist').columns({data:data.deployInputlist,
                    schema: [
                        {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadDeployInput(\'{{name}}\')">{{name}}</a>'},
                        {"header":"", "key":"cmd"},
                        {"header":" ", "key":"", "template":'<input type="button" onclick="deleteDeployInput(\'{{name}}\')" value="DELETE" />'},
                ]
                });
            }
            return
        }
        if(data.eventName == "dockerimagedeploy-load") {
            $("#name").val(data.deployInputData.name);
            $("#dit").val(data.deployInputData.dit);
            $("#port").val(data.deployInputData.port);
            $("#option").val(data.deployInputData.option);
            $("#dirname").val(data.deployInputData.dirname);
            $("#dirname2").val(data.deployInputData.dirname2);
            $("#dirname_a").val(data.deployInputData.dirname_a);
            $("#dirname_a2").val(data.deployInputData.dirname_a2);
            $("#dirname_b").val(data.deployInputData.dirname_b);
            $("#dirname_b2").val(data.deployInputData.dirname_b2);
            $("#dirname_c").val(data.deployInputData.dirname_c);
            $("#dirname_c2").val(data.deployInputData.dirname_c2);
            $("#dirname_d").val(data.deployInputData.dirname_d);
            $("#dirname_d2").val(data.deployInputData.dirname_d2);
            $("#option2").val(data.deployInputData.option2);
            //ないかもしれないのでラスト
            $("#image_id").val(data.deployInputData.image_id);
            $("#server_id").val(data.deployInputData.server_id);
            return
        }
      
        $.each(data.image, function (key, val) {
            $('#image_id').append($('<option>').val(val.ID).text(val.Repository+":"+val.Tag+":"+val.CreatedAt));
            console.log(val);
        });
        $.each(data.serverlist, function (key, val) {
            $('#server_id').append($('<option>').val(val.ip).text("server:"+val.user+"@"+val.ip));
        });
        $.each(data.vultr, function (key, val) {
            $('#server_id').append($('<option>').val(val.main_ip).text("vultr:"+key+":"+val.main_ip+"("+val.os+")"));
        });
        
        
        if(data.deployInputlist) {
            $('#inputlist').columns({data:data.deployInputlist,
                schema: [
                    {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadDeployInput(\'{{name}}\')">{{name}}</a>'},
                    {"header":"", "key":"cmd"},
                    {"header":" ", "key":"", "template":'<input type="button" onclick="deleteDeployInput(\'{{name}}\')" value="DELETE" />'},
            ]
            });
        }
    };

};
 
//エラーが発生した場合
ws.onerror = function(error) { };
 
//メッセージを受け取った場合
ws.onmessage = function(e) { };
  
//通信が切断された場合
ws.onclose = function() { };

//ページが読み込み完了した場合
$(document).ready(function () {
    $("#submit_button").click(function() {
        var fromData = formToData($("#frm"));
        fromData['event'] = "dockerimagedeploy-deploy"

        ws.send(JSON.stringify(fromData))
        console.log("Send success!!");
    });
    $("#submit_button2").click(function() {
        var fromData = formToData($("#frm"));
        fromData['event'] = "dockerimagedeploy-saveinput"

        ws.send(JSON.stringify(fromData))
        console.log("Send success!!");
    });
    $("#select_dir").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname").val(this.files[i].path)
          }
    });
    $("#select_dir2").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_a").val(this.files[i].path)
          }
    });
    $("#select_dir3").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_b").val(this.files[i].path)
          }
    });
    $("#select_dir4").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_c").val(this.files[i].path)
          }
    });
    $("#select_dir5").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_d").val(this.files[i].path)
          }
    });

});
function loadDeployInput(name) {

    ws.send(JSON.stringify({
        "event": "dockerimagedeploy-load",
        "name":name
    }));
}
function deleteDeployInput(name) {

    ws.send(JSON.stringify({
        "event": "dockerimagedeploy-delete",
        "name":name
    }));
}

File: mylib/socket/dockerimagedeploy/dockerimagedeploy.go
package dockerimagedeploy

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/uieviiru2/mylib/jsonedit"
	"github.com/uieviiru2/mylib/socket/config"
	"github.com/uieviiru2/mylib/socket/configserver"
	"github.com/uieviiru2/mylib/vultr"

	"github.com/Equanox/gotron"
	"github.com/uieviiru2/mylib/docker"
)

type DockerDeploy struct {
	ID        string `json:"image_id"`
	ServerIP  string `json:"server_ip"`
	Name      string `json:"name"`
	Dit       string `json:"dit"`
	Port      string `json:"port"`
	DirName   string `json:"dirname"`
	DirName2  string `json:"dirname2"`
	DirNameA  string `json:"dirname_a"`
	DirNameA2 string `json:"dirname_a2"`
	DirNameB  string `json:"dirname_b"`
	DirNameB2 string `json:"dirname_b2"`
	DirNameC  string `json:"dirname_c"`
	DirNameC2 string `json:"dirname_c2"`
	DirNameD  string `json:"dirname_d"`
	DirNameD2 string `json:"dirname_d2"`
	Option    string `json:"option"`
	Option2   string `json:"option2"`
	Cmd       string `json:"cmd"`
}

func Use(window *gotron.BrowserWindow) {

	window.On(&gotron.Event{Event: "dockerimagedeploy"}, func(bin []byte) {
		output := jsonedit.Val("eventName", "dockerimagedeploy")

		jDockerImage := docker.Image(window)
		output = jsonedit.Con(output, jDockerImage)

		jServerfiles := docker.Serverfiles(window)
		output = jsonedit.Con(output, jServerfiles)

		jVultr := vultr.List(window)
		output = jsonedit.Con(output, jVultr)

		jRunInputfiles := docker.DeployInputfiles()
		output = jsonedit.Con(output, jRunInputfiles)

		window.Send(&gotron.Event{Event: jsonedit.End(output)})
	})
	window.On(&gotron.Event{Event: "dockerimagedeploy-deploy"}, func(bin []byte) {
		//var d dockerID
		b := []byte(bin)
		buf := bytes.NewBuffer(b)
		fmt.Println(buf)
		var d DockerDeploy
		if err := json.Unmarshal(b, &d); err != nil {

		}
		configData := config.LoadConfig()
		jServer := docker.LoadServerfile(d.ServerIP)
		var d2 configserver.ConfigServer
		if err2 := json.Unmarshal([]byte(jServer), &d2); err2 != nil {
			//なければvultr
			d2.Ip = d.ServerIP
			d2.ServerPem = configData.VultrPem
			d2.User = "root"
		}
		go docker.ServerDeploy(d.ID, d.Name, d.Dit, d2.ServerPem, d2.User, d.ServerIP, d.Port, d.DirName, d.DirName2, d.DirNameA, d.DirNameA2, d.DirNameB, d.DirNameB2, d.DirNameC, d.DirNameC2, d.DirNameD, d.DirNameD2, d.Option, d.Option2, window)
		//fmt.Println(output)
		//window.Send(&gotron.Event{Event: jsonedit.End(output)})
	})
	window.On(&gotron.Event{Event: "dockerimagedeploy-saveinput"}, func(bin []byte) {
		var d DockerDeploy

		b := []byte(bin)
		//buf := bytes.NewBuffer(b)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(string(b))
		fmt.Println(d.ServerIP)
		fmt.Println(d.Name)
		fmt.Println(d.Port)
		fmt.Println(d.DirName)
		fmt.Println(d.DirName2)
		fmt.Println(d.DirNameA)
		fmt.Println(d.DirNameA2)
		fmt.Println(d.DirNameB)
		fmt.Println(d.DirNameB2)
		fmt.Println(d.DirNameC)
		fmt.Println(d.DirNameC2)
		fmt.Println(d.DirNameD)
		fmt.Println(d.DirNameD2)
		fmt.Println(d.Option)
		d.Cmd = makeRunStr(d)

		bytes, _ := json.Marshal(d)
		docker.SaveDeployInputfile(d.Name, string(bytes), window)

		output := jsonedit.Val("eventName", "dockerimagedeploy-saveinput")
		jRunInputfiles := docker.DeployInputfiles()
		output = jsonedit.Con(output, jRunInputfiles)
		window.Send(&gotron.Event{Event: jsonedit.End(output)})

	})
	window.On(&gotron.Event{Event: "dockerimagedeploy-load"}, func(bin []byte) {
		var d DockerDeploy
		b := []byte(bin)
		buf := bytes.NewBuffer(b)
		fmt.Println(buf)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(d.Name)

		output := jsonedit.Val("eventName", "dockerimagedeploy-load")
		jServer := docker.LoadDeployInputfile(d.Name)
		output = jsonedit.Con(output, jsonedit.On("deployInputData", jServer))

		window.Send(&gotron.Event{Event: jsonedit.End(output)})
	})
	window.On(&gotron.Event{Event: "dockerimagedeploy-delete"}, func(bin []byte) {
		var d DockerDeploy
		b := []byte(bin)
		buf := bytes.NewBuffer(b)
		fmt.Println(buf)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(d.Name)
		docker.DeleteDeployInputfile(d.Name, window)

		output := jsonedit.Val("eventName", "dockerimagedeploy-delete")
		jRunInputfiles := docker.DeployInputfiles()
		output = jsonedit.Con(output, jRunInputfiles)
		fmt.Println(output)
		window.Send(&gotron.Event{Event: jsonedit.End(output)})
	})
}

func makeRunStr(d DockerDeploy) string {
	cmd := "HOST:" + d.ServerIP + " docker run " + d.Dit + " --name " + d.Name
	if d.DirName != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirName) + ":/" + d.DirName2
	}
	if d.DirNameA != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameA) + ":/" + d.DirNameA2
	}
	if d.DirNameB != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameB) + ":/" + d.DirNameB2
	}
	if d.DirNameC != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameC) + ":/" + d.DirNameC2
	}
	if d.DirNameD != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameD) + ":/" + d.DirNameD2
	}
	if d.Port != "" {
		cmd += " -p " + d.Port
	}
	if d.Option != "" {
		cmd += " " + d.Option
	}
	cmd += " " + d.ID

	if d.Option2 != "" {
		cmd += " " + d.Option2
	}
	return cmd
}
